File: src/types/notifications.ts
export interface AgentMessageDelta {
  threadId: string;
  turnId: string;
  itemId: string;
  delta: string;
}

export interface TurnStartedNotification {
  threadId: string;
  turn: { id: string };
}

export interface TurnCompletedNotification {
  threadId: string;
  turn: { id: string };
}

export interface ItemStartedNotification {
  threadId: string;
  turnId: string;
  itemId: string;
  type: string;
}

export interface ItemCompletedNotification {
  threadId: string;
  turnId: string;
  itemId: string;
  type: string;
}

export interface CommandOutputDelta {
  threadId: string;
  turnId: string;
  itemId: string;
  delta: string;
}

export interface FileChangeOutputDelta {
  threadId: string;
  turnId: string;
  itemId: string;
  delta: string;
}

export interface PatchProgressChange {
  path: string;
  action: string;
  moveTo?: string;
}

export interface FileChangePatchUpdated {
  threadId?: string;
  turnId?: string;
  itemId?: string;
  callId?: string;
  path?: string;
  patch?: string;
  changes?: PatchProgressChange[];
}

export interface FileReviewReady {
  threadId?: string;
  itemId?: string;
  callId?: string;
  fileCount?: number;
}

export interface FileReviewUpdated {
  threadId?: string;
  callId?: string;
  status?: "updated" | "applied" | "cancelled" | "failed";
  message?: string;
  changedFiles?: Array<{ path: string; action: string }>;
}

export interface DocumentDetailInsertSnippet {
  snippet?: string;
}

export interface FortuneDetailStartedNotification {
  requestId: string;
}

export interface FortuneDetailDeltaNotification {
  requestId: string;
  delta: string;
}

export interface FortuneDetailCompletedNotification {
  requestId: string;
  text: string;
  finishReason?: string;
}

export interface FortuneDetailErrorNotification {
  requestId: string;
  message: string;
}

export interface ThreadStartedNotification {
  thread: {
    id: string;
    name?: string | null;
    preview?: string | null;
    archived?: boolean | null;
    updatedAt?: number | null;
  };
}

export interface ThreadStatusChangedNotification {
  threadId: string;
  status: string;
}

export interface ThreadNameUpdatedNotification {
  threadId: string;
  threadName?: string;
}

export interface ThreadTokenUsageUpdatedNotification {
  threadId: string;
  usage?: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
    cachedTokens?: number;
    cacheCreationTokens?: number;
    reasoningTokens?: number;
    callCount?: number;
    lastSinglePromptTokens?: number;
    contextWindowTokens?: number;
  };
  inputTokens?: number;
  outputTokens?: number;
  totalTokens?: number;
  cachedTokens?: number;
  cacheCreationTokens?: number;
  reasoningTokens?: number;
  callCount?: number;
  lastSinglePromptTokens?: number;
  contextPromptTokens?: number;
  modelContextWindow?: number;
}

export interface AccountUpdatedNotification {
  account: import("./account").Account;
}

export interface ErrorNotification {
  message: string;
  code?: string;
}

export interface ServerRequestEvent {
  requestId: import("./approval").RequestId;
  method: string;
  params: Record<string, unknown>;
}

export type ServerEventName =
  | "agent-message-delta"
  | "turn-started"
  | "turn-completed"
  | "turn-diff-updated"
  | "turn-plan-updated"
  | "item-started"
  | "item-completed"
  | "command-output-delta"
  | "file-change-output-delta"
  | "file-change-patch-updated"
  | "file-review-ready"
  | "file-review-updated"
  | "document-detail-insert-snippet"
  | "fortune-detail-started"
  | "fortune-detail-delta"
  | "fortune-detail-completed"
  | "fortune-detail-error"
  | "reasoning-text-delta"
  | "reasoning-summary-delta"
  | "plan-delta"
  | "hook-started"
  | "hook-completed"
  | "thread-started"
  | "thread-status-changed"
  | "thread-name-updated"
  | "thread-settings-updated"
  | "thread-goal-updated"
  | "thread-goal-cleared"
  | "thread-token-usage-updated"
  | "thread-archived"
  | "thread-unarchived"
  | "thread-closed"
  | "context-compacted"
  | "guardian-review-started"
  | "guardian-review-completed"
  | "account-updated"
  | "account-rate-limits-updated"
  | "account-login-completed"
  | "model-rerouted"
  | "model-verification"
  | "mcp-tool-call-progress"
  | "mcp-server-status-updated"
  | "server-error"
  | "server-warning"
  | "config-warning"
  | "skills-changed"
  | "server-request-resolved"
  | "server-request"
  | "events-lagged"
  | "server-notification";

File: src-tauri/src/protocol/notifications.rs
/// Server notification event names and dispatcher.
/// Instead of a large typed enum (like codex uses internally), cn-codex receives
/// JSON-RPC notifications from the external app-server process as raw JSON.
/// We match on the method name string and forward the params to Tauri events.

/// Maps a JSON-RPC notification method name to a Tauri event name.
pub fn notification_event_name(method: &str) -> &str {
    match method {
        // Turn lifecycle
        "turn/started" => "turn-started",
        "turn/completed" => "turn-completed",
        "turn/diff/updated" => "turn-diff-updated",
        "turn/plan/updated" => "turn-plan-updated",

        // Item lifecycle
        "item/started" => "item-started",
        "item/completed" => "item-completed",

        // Streaming deltas
        "item/agentMessage/delta" => "agent-message-delta",
        "item/reasoning/textDelta" => "reasoning-text-delta",
        "item/reasoning/summaryTextDelta" => "reasoning-summary-delta",
        "item/reasoning/summaryPartAdded" => "reasoning-summary-part-added",
        "item/plan/delta" => "plan-delta",

        // Command execution
        "command/exec/outputDelta" => "command-output-delta",
        "item/commandExecution/outputDelta" => "command-output-delta",

        // File changes
        "item/fileChange/outputDelta" => "file-change-output-delta",
        "item/fileChange/patchUpdated" => "file-change-patch-updated",

        // Hooks
        "hook/started" => "hook-started",
        "hook/completed" => "hook-completed",

        // Thread lifecycle
        "thread/started" => "thread-started",
        "thread/status/changed" => "thread-status-changed",
        "thread/name/updated" => "thread-name-updated",
        "thread/settingsUpdated" => "thread-settings-updated",
        "thread/goalUpdated" => "thread-goal-updated",
        "thread/goalCleared" => "thread-goal-cleared",
        "thread/tokenUsage/updated" => "thread-token-usage-updated",
        "thread/compacted" => "context-compacted",
        "thread/archived" => "thread-archived",
        "thread/unarchived" => "thread-unarchived",
        "thread/closed" => "thread-closed",

        // Guardian / auto-approval
        "item/autoApprovalReview/started" => "guardian-review-started",
        "item/autoApprovalReview/completed" => "guardian-review-completed",

        // Account
        "account/updated" => "account-updated",
        "account/rateLimits/updated" => "account-rate-limits-updated",
        "account/login/completed" => "account-login-completed",

        // Model
        "model/rerouted" => "model-rerouted",
        "model/verification" => "model-verification",

        // MCP
        "item/mcpToolCall/progress" => "mcp-tool-call-progress",
        "mcpServer/startupStatus/updated" => "mcp-server-status-updated",

        // Error/warning
        "error" => "server-error",
        "warning" => "server-warning",
        "configWarning" => "config-warning",

        // Skills
        "skills/changed" => "skills-changed",

        // Server request resolved
        "serverRequest/resolved" => "server-request-resolved",

        // Catch-all
        _ => "server-notification",
    }
}
